lib/graph_gateway.go

```diff
@@ -0,0 +1,482 @@
+}
+
+type gwOptions struct {
+	ns namesys.NameSystem
+	vs routing.ValueStore
+	bs blockstore.Blockstore
+}
+
+// WithNameSystem sets the name system to use for the gateway. If not set it will use a default DNSLink resolver
+// along with any configured ValueStore
+func WithNameSystem(ns namesys.NameSystem) GraphGatewayOption {
+	return func(opts *gwOptions) error {
+		opts.ns = ns
+		return nil
+	}
+}
+
+// WithValueStore sets the ValueStore to use for the gateway
+func WithValueStore(vs routing.ValueStore) GraphGatewayOption {
+	return func(opts *gwOptions) error {
+		opts.vs = vs
+		return nil
+	}
+}
+
+// WithBlockstore sets the Blockstore to use for the gateway
+func WithBlockstore(bs blockstore.Blockstore) GraphGatewayOption {
+	return func(opts *gwOptions) error {
+		opts.bs = bs
+		return nil
+	}
+}
+
+type GraphGatewayOption func(gwOptions *gwOptions) error
+
+type GraphGateway struct {
+	fetcher      CarFetcher
+	blockFetcher exchange.Fetcher
+	routing      routing.ValueStore
+	namesys      namesys.NameSystem
+	bstore       blockstore.Blockstore
+	bsrv         blockservice.BlockService
+}
+
+func NewGraphGatewayBackend(f CarFetcher, blockFetcher exchange.Fetcher, opts ...GraphGatewayOption) (*GraphGateway, error) {
+	var compiledOptions gwOptions
+	for _, o := range opts {
+		if err := o(&compiledOptions); err != nil {
+			return nil, err
+		}
+	}
+
+	// Setup a name system so that we are able to resolve /ipns links.
+	vs := compiledOptions.vs
+	if vs == nil {
+		vs = routinghelpers.Null{}
+	}
+
+	ns := compiledOptions.ns
+	if ns == nil {
+		dns, err := gateway.NewDNSResolver(nil, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		ns, err = namesys.NewNameSystem(vs, namesys.WithDNSResolver(dns))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	bs := compiledOptions.bs
+	if compiledOptions.bs == nil {
+		// Sets up a cache to store blocks in
+		cbs, err := NewCacheBlockStore(DefaultCacheBlockStoreSize)
+		if err != nil {
+			return nil, err
+		}
+
+		// Set up support for identity hashes (https://github.com/ipfs/bifrost-gateway/issues/38)
+		cbs = blockstore.NewIdStore(cbs)
+		bs = cbs
+	}
+
+	return &GraphGateway{
+		fetcher:      f,
+		blockFetcher: blockFetcher,
+		routing:      vs,
+		namesys:      ns,
+		bstore:       bs,
+	}, nil
+}
+
+/*
+Implementation iteration plan:
+
+1. Fetch CAR into per-request memory blockstore and serve response
+2. Fetch CAR into shared memory blockstore and serve response along with a blockservice that does block requests for missing data
+3. Start doing the walk locally and then if a path segment is incomplete send a request for blocks and upon every received block try to continue
+4. Start doing the walk locally and keep a list of "plausible" blocks, if after issuing a request we get a non-plausible block then report them and attempt to recover by redoing the last segment
+5. Don't redo the last segment fully if it's part of a UnixFS file and we can do range requests
+*/
```

Comment:
Calling out that this is the basic plan. At the moment 1 is done. 2 is mostly done. We don't need to land all of this before merging or doing testing with mirror traffic but we do need 4 for safety DoS purposes when being used with actually untrusted backends.

Reply:
:thought_balloon:  @aschmahmann  when we start doing the walk locally, and only fetch missing parts of graph, caboose will be able to see that the requested path is different than the [original Content Path requested by the user that we pass in context](https://github.com/filecoin-saturn/caboose/blob/e621c985ff5baf4f229ebd949b9916e2f8b8f235/caboose.go#L273).

My thinking is that for such "internal" block/CAR requests, we will want to forward the parent content path to the backend HTTP server  in `Ipfs-Path-Affinity` header. 

It does not cost us much, but will allow for caching/content routing optimizations (e.g. if the root of CAR i requested is not on DHT, Lassie will still be able to find it because it was able to find root CID from  `Ipfs-Path-Affinity`, if present).